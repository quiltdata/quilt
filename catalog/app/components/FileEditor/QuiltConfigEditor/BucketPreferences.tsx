import * as React from 'react'
import * as M from '@material-ui/core'

import bucketPreferencesSchema from 'schemas/bucketConfig.yml.json'

import { docs } from 'constants/urls'
import StyledLink from 'utils/StyledLink'

import { ConfigDetailsProps } from './Dummy'

function Header() {
  return (
    <M.Typography variant="body2">
      Configuration for Catalog UI: show and hide features, set default values. See{' '}
      <StyledLink
        href={`${docs}/quilt-platform-administrator/preferences`}
        target="_blank"
      >
        the docs
      </StyledLink>
    </M.Typography>
  )
}

export default function BucketPreferences({ children }: ConfigDetailsProps) {
  return children({ header: <Header />, schema: bucketPreferencesSchema })
}
