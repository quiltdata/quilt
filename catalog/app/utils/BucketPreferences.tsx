import * as R from 'ramda'
import * as React from 'react'

import bucketPreferencesSchema from 'schemas/bucketConfig.yml.json'

import * as AWS from 'utils/AWS'
import { useData } from 'utils/Data'
import * as Sentry from 'utils/Sentry'
import { makeSchemaValidator } from 'utils/json-schema'
import yaml from 'utils/yaml'
import * as bucketErrors from 'containers/Bucket/errors'
import * as requests from 'containers/Bucket/requests'

type SentryInstance = (command: 'captureMessage', message: string) => void

export type ActionPreferences = Record<
  'copyPackage' | 'createPackage' | 'deleteRevision' | 'revisePackage',
  boolean
>

export type NavPreferences = Record<'files' | 'packages' | 'queries', boolean>

export interface SourceBuckets {
  getDefault: () => string
  list: string[]
}

interface UiPreferences {
  actions: ActionPreferences
  nav: NavPreferences
  sourceBuckets: SourceBuckets
}

interface UiPreferencesYaml {
  actions?: ActionPreferences
  defaultSourceBucket?: string
  nav?: NavPreferences
  sourceBuckets?: Record<string, null>
}

interface BucketPreferencesYaml {
  ui?: UiPreferencesYaml
}

interface BucketPreferences {
  ui: UiPreferences
}

const defaultPreferences: BucketPreferences = {
  ui: {
    actions: {
      copyPackage: true,
      createPackage: true,
      deleteRevision: false,
      revisePackage: true,
    },
    nav: {
      files: true,
      packages: true,
      queries: true,
    },
    sourceBuckets: {
      getDefault: () => '',
      list: [],
    },
  },
}

const bucketPreferencesValidator = makeSchemaValidator(bucketPreferencesSchema)

function validate(data: unknown): asserts data is BucketPreferencesYaml {
  const errors = bucketPreferencesValidator(data)
  if (errors.length) throw new bucketErrors.BucketPreferencesInvalid({ errors })
}

const S3_PREFIX = 's3://'
const normalizeBucketName = (input: string) =>
  input.startsWith(S3_PREFIX) ? input.slice(S3_PREFIX.length) : input

function parseSourceBuckets(
  sentry: SentryInstance,
  ui?: UiPreferencesYaml,
): SourceBuckets {
  const list = Object.keys(ui?.sourceBuckets || {}).map(normalizeBucketName)
  const defaultSourceBucket = normalizeBucketName(ui?.defaultSourceBucket || '')
  return {
    getDefault: () => {
      if (defaultSourceBucket) {
        const found = list.find((name) => name === defaultSourceBucket)
        if (found) return found
        sentry(
          'captureMessage',
          `defaultSourceBucket ${defaultSourceBucket} is incorrect`,
        )
      }
      return list[0] || ''
    },
    list,
  }
}

function parse(bucketPreferencesYaml: string, sentry: SentryInstance): BucketPreferences {
  const data = yaml(bucketPreferencesYaml)
  if (!data) return defaultPreferences

  validate(data)

  return {
    ui: {
      actions: R.mergeRight(defaultPreferences.ui.actions, data?.ui?.actions || {}),
      nav: R.mergeRight(defaultPreferences.ui.nav, data?.ui?.nav || {}),
      sourceBuckets: parseSourceBuckets(sentry, data?.ui),
    },
  }
}

const BUCKET_PREFERENCES_PATH = [
  '.quilt/catalog/config.yaml',
  '.quilt/catalog/config.yml',
]

interface FetchBucketPreferencesArgs {
  s3: $TSFixMe
  bucket: string
  sentry: SentryInstance
}

const fetchBucketPreferences = async ({
  s3,
  sentry,
  bucket,
}: FetchBucketPreferencesArgs) => {
  try {
    const response = await requests.fetchFile({
      s3,
      bucket,
      path: BUCKET_PREFERENCES_PATH,
    })
    return parse(response.Body.toString('utf-8'), sentry)
  } catch (e) {
    if (
      e instanceof bucketErrors.FileNotFound ||
      e instanceof bucketErrors.VersionNotFound
    )
      return defaultPreferences

    // eslint-disable-next-line no-console
    console.log('Unable to fetch')
    // eslint-disable-next-line no-console
    console.error(e)
    throw e
  }
}

const Ctx = React.createContext<BucketPreferences | null>(null)

type ProviderProps = React.PropsWithChildren<{ bucket: string }>

export function Provider({ bucket, children }: ProviderProps) {
  const sentry = Sentry.use()
  const s3 = AWS.S3.use()
  const data = useData(fetchBucketPreferences, { s3, sentry, bucket })

  // XXX: consider returning AsyncResult or using Suspense
  const preferences = data.case({
    Ok: R.identity,
    Err: () => defaultPreferences,
    _: () => null,
  })
  return <Ctx.Provider value={preferences}>{children}</Ctx.Provider>
}

export const useBucketPreferences = () => React.useContext(Ctx)

export const use = useBucketPreferences
