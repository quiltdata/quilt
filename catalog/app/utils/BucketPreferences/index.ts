export type {
  BucketPreferences,
  NavPreferences,
  SourceBuckets,
} from './BucketPreferences'
export * from './Provider'
