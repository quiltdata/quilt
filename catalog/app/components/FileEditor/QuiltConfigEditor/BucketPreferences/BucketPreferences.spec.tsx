import * as React from 'react'
import renderer from 'react-test-renderer'
import { createMuiTheme } from '@material-ui/core'

import BucketPreferences from './BucketPreferences'

const theme = createMuiTheme()
const noop = () => {}

jest.mock(
  'constants/config',
  jest.fn(() => ({})),
)

jest.mock(
  'utils/BucketConfig',
  jest.fn(() => ({
    useRelevantBucketConfigs: () => [],
  })),
)

jest.mock(
  '@material-ui/core',
  jest.fn(() => ({
    ...jest.requireActual('@material-ui/core'),
    Checkbox: jest.fn(({ checked }: { checked: boolean }) => (
      <div id="checkbox">{checked.toString()}</div>
    )),
    IconButton: jest.fn(({ children }: { children: React.ReactNode }) => (
      <div id="icon-button">{children}</div>
    )),
    Icon: jest.fn(({ children }: { children: string }) => (
      <div id="icon">{children}</div>
    )),
    TextField: jest.fn(({ value }: { value: React.ReactNode }) => (
      <div id="text-field">{value}</div>
    )),
    Select: jest.fn(({ value }: { value: string }) => <div id="select">{value}</div>),
    makeStyles: jest.fn((cb: any) => () => {
      const classes = typeof cb === 'function' ? cb(theme) : cb
      return Object.keys(classes).reduce(
        (acc, key) => ({
          [key]: key,
          ...acc,
        }),
        {},
      )
    }),
  })),
)

jest.mock(
  '@material-ui/lab',
  jest.fn(() => ({
    ...jest.requireActual('@material-ui/core'),
    Autocomplete: jest.fn(({ options, value }: { options?: string[]; value: string }) => (
      <div id="autocomplete">
        {value} from [{options?.join(', ')}]
      </div>
    )),
  })),
)

describe('components/FileEditor/QuiltConfigEditor/BucketPreferences/BucketPreferences', () => {
  it('render form with default values', () => {
    const tree = renderer
      .create(<BucketPreferences className="root" error={null} onChange={noop} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('render form config where every value is custom', () => {
    const config = `ui:
  actions:
    copyPackage: false
    createPackage: false
    deleteRevision: true
    downloadObject: false
    downloadPackage: false
    openInDesktop: true
    revisePackage: false
    writeFile: false
  blocks:
    analytics: false
    browser: false
    code: false
    meta:
      user_meta:
        expanded: 2
      workflows:
        expanded: 3
    gallery:
      files: false
      overview: false
      packages: false
      summarize: false
    qurator: false
  nav:
    files: false
    packages: false
    queries: false
  sourceBuckets:
    s3://a: {}
    s3://b: {}
    s3://c: {}
  defaultSourceBucket: s3://b
  package_description:
    ^prefix/namespace$:
      message: false
      user_meta:
        - $.Some.Key
        - $.Another.Key
  package_description_multiline: true
  athena:
    defaultWorkgroup: Foo
`
    const tree = renderer
      .create(
        <BucketPreferences
          initialValue={config}
          className="root"
          error={null}
          onChange={noop}
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
