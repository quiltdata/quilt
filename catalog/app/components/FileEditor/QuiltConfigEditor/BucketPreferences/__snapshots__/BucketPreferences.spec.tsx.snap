// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FileEditor/QuiltConfigEditor/BucketPreferences/BucketPreferences render form config where every value is custom 1`] = `
<div
  className="root root"
>
  <p
    className="MuiTypography-root helperText MuiTypography-body2"
  >
    Per-bucket Catalog UI configuration: show and hide features, set default values.
     
    <a
      className="root"
      href="https://docs.quiltdata.com/quilt-platform-administrator/preferences"
      target="_blank"
    >
      Learn more
    </a>
  </p>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Actions
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Show buttons and menu items
      </p>
    </h6>
    <div
      className="layout duplex"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "PUSH TO BUCKET" on the package page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "CREATE PACKAGE" on the package list and bucket listing pages
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "DELETE REVISION" menu item on the package page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Download buttons under the "BUCKET" tab
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Download buttons under the "PACKAGES" tab
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "REVISE PACKAGE" on the package page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Buttons to create or edit files
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "Qurator Assistance" icon button on object and package pages
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Sections
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Show UI sections
      </p>
    </h6>
    <div
      className="layout duplex"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "ANALYTICS" on the file page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            File listings on bucket and packages pages
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "CODE"
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "METADATA"
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Metadata Display
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Display settings for JSON blocks in Metadata section on object and package pages
      </p>
    </h6>
    <div
      className="layout"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Auto-expand "User Metadata" field
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Auto-expand "Workflow" field
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Galleries
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Show image galleries
      </p>
    </h6>
    <div
      className="layout duplex"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Images in the directory listing on bucket pages
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Images on the Bucket overview page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Images in the directory listing on the package page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Image galleries alongside those defined in quilt_summarize.json
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Navigation items
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Show tabs at the top of the bucket pages
      </p>
    </h6>
    <div
      className="layout"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "BUCKET"
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "PACKAGES"
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "QUERIES"
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Source buckets for packages
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Buckets available in package creation and revision dialogs under "ADD FILES FROM BUCKET"
      </p>
    </h6>
    <div
      className="layout duplex"
    >
      <div
        id="autocomplete"
      >
        s3://a
        s3://b
        s3://c
         from [
        
        ]
      </div>
      <div
        className="MuiFormControl-root margin MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
          data-shrink={false}
        >
          Default bucket
        </label>
        <div
          id="select"
        >
          s3://b
        </div>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Package List: common display settings
      
    </h6>
    <div
      className="layout"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Display \`user_meta\` fields on separate lines, when enabled. Visibility of the \`user_meta\` can be configured in the next section.
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Package List: selective display settings
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Selectively apply display settings to matching packages
      </p>
    </h6>
    <div
      className="layout"
    >
      <div
        className="root margin"
      >
        <div
          className="root"
        >
          <div
            id="text-field"
          >
            ^prefix/namespace$
          </div>
          <div
            id="autocomplete"
          >
            $.Some.Key
            $.Another.Key
             from [
            
            ]
          </div>
          <div
            className="MuiFormControl-root"
          >
            <label
              className="MuiFormControlLabel-root"
            >
              <div
                id="checkbox"
              >
                false
              </div>
              <span
                className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
              >
                Display commit message
              </span>
            </label>
          </div>
          <div
            id="icon-button"
          >
            <div
              id="icon"
            >
              close
            </div>
          </div>
        </div>
        <div
          className="root"
        >
          <div
            id="text-field"
          >
            
          </div>
          <div
            id="autocomplete"
          >
             from [
            
            ]
          </div>
          <div
            className="MuiFormControl-root"
          >
            <label
              className="MuiFormControlLabel-root Mui-disabled"
            >
              <div
                id="checkbox"
              >
                false
              </div>
              <span
                className="MuiTypography-root MuiFormControlLabel-label Mui-disabled MuiTypography-body1"
              >
                Display commit message
              </span>
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Athena
      
    </h6>
    <div
      className="layout"
    >
      <div
        id="text-field"
      >
        Foo
      </div>
    </div>
  </div>
</div>
`;

exports[`components/FileEditor/QuiltConfigEditor/BucketPreferences/BucketPreferences render form with default values 1`] = `
<div
  className="root root"
>
  <p
    className="MuiTypography-root helperText MuiTypography-body2"
  >
    Per-bucket Catalog UI configuration: show and hide features, set default values.
     
    <a
      className="root"
      href="https://docs.quiltdata.com/quilt-platform-administrator/preferences"
      target="_blank"
    >
      Learn more
    </a>
  </p>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Actions
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Show buttons and menu items
      </p>
    </h6>
    <div
      className="layout duplex"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "PUSH TO BUCKET" on the package page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "CREATE PACKAGE" on the package list and bucket listing pages
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "DELETE REVISION" menu item on the package page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Download buttons under the "BUCKET" tab
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Download buttons under the "PACKAGES" tab
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "REVISE PACKAGE" on the package page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Buttons to create or edit files
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "Qurator Assistance" icon button on object and package pages
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Sections
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Show UI sections
      </p>
    </h6>
    <div
      className="layout duplex"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "ANALYTICS" on the file page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            File listings on bucket and packages pages
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "CODE"
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "METADATA"
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Metadata Display
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Display settings for JSON blocks in Metadata section on object and package pages
      </p>
    </h6>
    <div
      className="layout"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Auto-expand "User Metadata" field
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Auto-expand "Workflow" field
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Galleries
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Show image galleries
      </p>
    </h6>
    <div
      className="layout duplex"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Images in the directory listing on bucket pages
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Images on the Bucket overview page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Images in the directory listing on the package page
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Image galleries alongside those defined in quilt_summarize.json
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Navigation items
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Show tabs at the top of the bucket pages
      </p>
    </h6>
    <div
      className="layout"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "BUCKET"
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "PACKAGES"
          </span>
        </label>
      </div>
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            true
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            "QUERIES"
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Source buckets for packages
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Buckets available in package creation and revision dialogs under "ADD FILES FROM BUCKET"
      </p>
    </h6>
    <div
      className="layout duplex"
    >
      <div
        id="autocomplete"
      >
         from [
        
        ]
      </div>
      <div
        className="MuiFormControl-root margin MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense Mui-disabled Mui-disabled"
          data-shrink={false}
        >
          Default bucket
        </label>
        <div
          id="select"
        >
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Package List: common display settings
      
    </h6>
    <div
      className="layout"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <div
            id="checkbox"
          >
            false
          </div>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Display \`user_meta\` fields on separate lines, when enabled. Visibility of the \`user_meta\` can be configured in the next section.
          </span>
        </label>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Package List: selective display settings
      <p
        className="MuiTypography-root description MuiTypography-body2"
      >
        Selectively apply display settings to matching packages
      </p>
    </h6>
    <div
      className="layout"
    >
      <div
        className="root margin"
      >
        <div
          className="root"
        >
          <div
            id="text-field"
          >
            .*
          </div>
          <div
            id="autocomplete"
          >
             from [
            
            ]
          </div>
          <div
            className="MuiFormControl-root"
          >
            <label
              className="MuiFormControlLabel-root"
            >
              <div
                id="checkbox"
              >
                true
              </div>
              <span
                className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
              >
                Display commit message
              </span>
            </label>
          </div>
          <div
            id="icon-button"
          >
            <div
              id="icon"
            >
              close
            </div>
          </div>
        </div>
        <div
          className="root"
        >
          <div
            id="text-field"
          >
            
          </div>
          <div
            id="autocomplete"
          >
             from [
            
            ]
          </div>
          <div
            className="MuiFormControl-root"
          >
            <label
              className="MuiFormControlLabel-root Mui-disabled"
            >
              <div
                id="checkbox"
              >
                false
              </div>
              <span
                className="MuiTypography-root MuiFormControlLabel-label Mui-disabled MuiTypography-body1"
              >
                Display commit message
              </span>
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="group"
  >
    <h6
      className="MuiTypography-root title MuiTypography-h6"
    >
      Athena
      
    </h6>
    <div
      className="layout"
    >
      <div
        id="text-field"
      >
        
      </div>
    </div>
  </div>
</div>
`;
